src/ssl/support.cc

```diff
@@ -423,6 +435,77 @@ Ssl::DisablePeerVerification(Security::ContextPointer &ctx)
 }
 
+static int squidX509VerifyCert(X509_STORE_CTX *ctx, STACK_OF(X509) *extraCerts);
+
+bool
+Ssl::PeerCertificatesVerify(Security::SessionPointer &s, const Ssl::X509_STACK_Pointer &extraCerts)
+{
+    STACK_OF(X509) *peerCertificatesChain = SSL_get_peer_cert_chain(s.get());
+
+    if (!peerCertificatesChain || sk_X509_num(peerCertificatesChain) == 0) {
+        debugs(83, 2, "no server certificates?");
+        return false;
+    }
+
+    X509_STORE *verificationStore = SSL_CTX_get_cert_store(SSL_get_SSL_CTX(s.get()));
+    X509_STORE_CTX_Pointer storeCtx(X509_STORE_CTX_new());
+    X509 *peerCert = sk_X509_value(peerCertificatesChain, 0);
+    if (!X509_STORE_CTX_init(storeCtx.get(), verificationStore, peerCert, peerCertificatesChain)) {
+        debugs(83, DBG_IMPORTANT, "error initializing X509_STORE");
+        return false;
+    }
+
+    // Retrieve certificate flags:
+    long certFlags = SSL_set_cert_flags(s.get(), 0);
+    if (certFlags & SSL_CERT_FLAG_SUITEB_128_LOS) {
+        auto SSL_CERT_FLAG_to_X509_V_FLAG = [](long flag) {return flag;};
+        long x509Flags = SSL_CERT_FLAG_to_X509_V_FLAG(certFlags);
+        X509_STORE_CTX_set_flags(storeCtx.get(), x509Flags & X509_V_FLAG_SUITEB_128_LOS);
+    }
+
+    if (!X509_STORE_CTX_set_ex_data(storeCtx.get(), SSL_get_ex_data_X509_STORE_CTX_idx(), s.get())) {
+        debugs(83, DBG_IMPORTANT, "error attaching session object to X509_STORE");
+        return false;
+    }
+
+    // The X509_verify_cert will only use dane if it is enabled.
+    X509_STORE_CTX_set0_dane(storeCtx.get(), SSL_get0_dane(s.get()));
+
+    // We are verifying a server certificate
+    X509_STORE_CTX_set_default(storeCtx.get(), "ssl_server");
+
+    X509_VERIFY_PARAM *param = X509_STORE_CTX_get0_param(storeCtx.get());
+    assert(param);
+    X509_VERIFY_PARAM_set_auth_level(param, SSL_get_security_level(s.get()));
+    X509_VERIFY_PARAM_set1(param, SSL_get0_param(s.get()));
+
+    if (auto cb = SSL_get_verify_callback(s.get()))
+        X509_STORE_CTX_set_verify_cb(storeCtx.get(), cb);
+
+    // Squid implements its own verify function to use instead of
+    // using X509_verify_cert, which also adds missing certificates
+    // to storeCtx before checks.
+    // bool verified = (X509_verify_cert(storeCtx.get()) > 0);
+    bool verified = (squidX509VerifyCert(storeCtx.get(), extraCerts.get()) > 0);
+
+    if (!verified) {
+        // Our call back will record errors list
+        int verifyResult = X509_STORE_CTX_get_error(storeCtx.get());
+        debugs(83, 5, "verify result " << verifyResult << " : " << X509_verify_cert_error_string(verifyResult));
+    }
+    return verified;
+}
+
+void Ssl::SquidVerifyData::SetSessionFlag(Security::SessionPointer &session, uint64_t flg)
+{
+    auto data = SessionData(session);
+    if (!data) {
+        data =  new Ssl::SquidVerifyData();
+        SSL_set_ex_data(session.get(), ssl_ex_index_squid_verify, data);
+    }
+    data->set(Ssl::SquidVerifyData::fIgnoreIssuer);
+}
```

Comment:
This function is buggy -- it does not use its `flg` parameter. That is easy to fix, but I do not think this function should exist at all in the current code. Here is why: AFAICT, this is the only case where we _set_ the ssl_ex_index_squid_verify ex_data. The function itself is only called from one place and is called unconditionally. All other places calling SessionData() have to assert that SessionData() returns a non-nil pointer. If we inline this function body into its caller, the code will become much simpler -- the `if` statement can be removed and all the remaining SessionData() callers will receive a reference to the "session data" object, not a pointer. If I am right, please adjust accordingly. Otherwise, please let me know what I missed.

Reply:
> This function is buggy -- it does not use its `flg` parameter. 

It is intended. To show you that currently we need only an on value on/off stored to  "ssl_ex_index_squid_verify" to ssl_ex_data index.

However we are also using the ssl_ex_index_server, ssl_ex_index_cert_error_check, ssl_ex_index_ssl_peeked_cert and  ssl_ex_index_ssl_validation_counter indexes to store various data used in verify procedure and not only. 

My plan is to merge all this info into the new Ssl::SquidVerifyData object attached to ssl_ex_index_squid_verify index, not into this issue but on a future project. 
For example the ssl_ctx_ex_index_dont_verify_domain ssl_ex index is a flag too, the flag code in SquidVerifyData can replace it.

Else we can drop Ssl::SquidVerifyData class and use only an Ssl::SetSessionFlag which sets an uint_t (and not a ) to ssl_ex_index_squid_verify index.

Reply:
> It is intended.

We should not leave unused named parameters, especially intentionally. Any rule of thumb has exceptions, of course, but I see no reason to trick the code reader in this particular case. 

N.B. In the future, if you are convinced that some new code should do something tricky like this, please add a corresponding source code comment to explicitly flag (and justify) the exceptional case!

> My plan is to merge ssl_ex_index_server, ssl_ex_index_cert_error_check, ssl_ex_index_ssl_peeked_cert and ssl_ex_index_ssl_validation_counter into the new Ssl::SquidVerifyData object attached to ssl_ex_index_squid_verify index.

Overall, that sounds like a good plan to me. I am not against this PR adding a class like SquidVerifyData to encapsulate certificate validation metadata. And I agree that we do not have to refactor the old code here, in this PR. None of that justifies having a SetSessionFlag() function and having SessionData() that returns a pointer (forcing all other callers to assert that it is not nil).

I doubt the future SquidVerifyData-like class will need a SetSessionFlag() function. I suspect the code will set flags/fields directly instead of using bitwise logic and enums. If I am wrong, we can always add SetSessionFlag() later, of course.

Reply:
Removed with commit 1582254825 and the related code refactored to use bools instead of bin flags with commit adf44d10ff